### ISSUE TITLE
Incorrect error on Invariants typed with Tuple containing Literal as first argument

### ISSUE THREAD

**Issue by SebbyLaw** [microsoft/pyright#3722]
Sorry if the title or wording is unclear or wrong. This is my first bug report on pyright.

**Describe the bug**

Invariant collection types typed as containing a Tuple of a Literal and a second item make pyright raise an incorrect error when the second item does not conform to the type of the second item in the Tuple.

The incorrect error 

**To Reproduce**

See the code below, using the latest version of pylance (`2022.7.21` (6 July 2022)).

**Expected behavior**

Pyright should point the user in the right direction by providing an accurate type error.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
from typing import Coroutine, List, Literal, Tuple

Method = Literal["GET", "POST", "PUT", "PATCH", "DELETE"]


async def coro():
    pass


# below definitions should error in pyright

# gives a weird error
bad: List[Tuple[Method, Coroutine]] = [
    ("DELETE", coro),
]

# gives the correct error when the first Tuple item is not typed as a Literal
good: List[Tuple[str, Coroutine]] = [
    ("DELETE", coro),
]
```

Error for `bad`:

![Code_ptmqNJasTF](https://user-images.githubusercontent.com/44045823/179863337-fb24b056-ccfb-4e6e-80f1-2f80dac55811.png)

Error for `good`:

![Code_GY3oAlY8Uw](https://user-images.githubusercontent.com/44045823/179863353-82d4fc4a-caf1-4da6-a6d9-3dafc9bc42e1.png)

**VS Code extension or command-line**

Running the latest version of pylance for VSCode in "basic" mode.
(pylance 2022.7.21, pyright 1.1.257)

**Comment by erictraut** [comment]
Pyright is working correctly here.

The problem has nothing to do with literals. Let's simplify your example, which might help you better understand what's going on here:
```python
example1: list[tuple[str | int, str]] = [("", 0)]
example2: list[tuple[str, str]] = [("", 0)]
```

Pyright attempts to use [bidirectional type inference](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#bidirectional-type-inference-expected-types) to infer the type of the right-hand side (RHS) of the assignment statement. In both cases (`example1` and `example2`), bidirectional type inference fails because the RHS expression is not compatible with the expected type (i.e. the declared type of the variable). Pyright therefore falls back on default inference rules for [lists](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#list-expressions) and [tuples](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#tuple-expressions). In both cases, the inferred type of the RHS is `list[tuple[Literal[''], Literal[0]]`. For both examples, the assigned type is incompatible with the declared type. For `example1`, both entries of the tuple are incompatible types, but pyright stops after determining that the first entry is incompatible. For `example2`, the first entry of the tuple is compatible, but the second entry of the tuple is incompatible.

That explains why you're seeing the error messages that you are seeing.

**Comment by SebbyLaw** [comment]
> For `example1`, both entries of the tuple are incompatible types, but pyright stops after determining that the first entry is incompatible. For `example2`, the first entry of the tuple is compatible, but the second entry of the tuple is incompatible.

Sorry, I don't see how this relates. The issue I was bringing up was the first entry of the tuple always being valid, but pyright saying it is invalid when in reality the second entry was invalid.

I've simplified my examples even more:
```py
example3: list[tuple[Literal["a", "b"], str]] = [("a", 0)]  # reports the first entry as invalid, even though it is valid
example4: list[tuple[Literal["a", "b"], str]] = [("b", 0)]  # reports the first entry as invalid, even though it is valid
example5: list[tuple[Literal["a"], str]] = [("a", 0)]  # reports the second entry as invalid, as it should
example6: list[tuple[Literal["a"], str]] = [("b", 0)]  # reports the first entry as invalid, as it should
example7: list[tuple[Literal["a", "b"], str]] = [("a", "foo")]  # both entries are valid, works as intended
example8: list[tuple[Literal["a", "b"], str]] = [("b", "bar")]  # both entries are valid, works as intended
```
This bug seems to only happen when the first entry is a Literal with multiple possibilities

**Comment by erictraut** [comment]
When you say "valid", I presume that you mean "type compatible". The first entry in the tuple is not type compatible as you are contending. That's because the tuple is enclosed in a list. The type parameter for `list` is invariant, and that means that type enclosed within the list must match exactly.

The type parameter for `tuple` is covariant, so subtypes are compatible.

```python
def test1(val: tuple[int, int]):
    x: tuple[int | str, int] = val # No type violation
```

The type parameter for `list` is invariant, so the types need to match exactly.

```python
def test2(val: list[int]):
    x: list[int | str] = val # Type violation
```

The same invariance rules hold for a tuple once you enclose it in a list.

```python
def test3(val: list[tuple[int, int]]):
    x: list[tuple[int | str, int]] = val # Type violation
```

So this is not a bug in pyright. It's correctly identifying the type violation and telling you that it is because the type parameter for `list` is invariant.

**Comment by Rapptz** [comment]
I don't think the problem is variance here. It just seems like a bug in the error reporting.

Using the previous example3 and example4, the error shows both types being incompatible:

```py
example3: list[tuple[Literal["a", "b"], str]] = [("a", 0)]
```

![image](https://user-images.githubusercontent.com/1695103/179900393-ebe13747-b8d8-426d-8d6c-c9dc6e630a24.png)

Where as something like 

```py
example7: list[tuple[Literal["a", "b"], str]] = [("a" ,"foo")]
```

Would pass, implying that the error is pointing to the wrong direction since the first element of the tuple is fine (`Literal["a"]` can be assigned to `Literal["a', "b"]`).

Using an enum also causes the error to be fine:

![image](https://user-images.githubusercontent.com/1695103/179900697-2067031c-c799-461f-a525-7f1ccb5c84dc.png)

**Comment by erictraut** [comment]
In `example7`, the RHS expression's type is inferred as `list[tuple[Literal["a", "b"], str]]` via bidirectional type inference. But in `example3`, the bidirectional type inference fails, so default inference rules apply, and the inferred type is `list[tuple[Literal["a"], Literal[0]]]`. In this case, the first entry of the tuple (of type `Literal["a"]`) is not compatible with `Literal["a", "b"]` in the declared type because of invariance.

Note that `Literal["a", "b"]` is simply shorthand for the union `Literal["a"] | Literal["b"]`. If you change your `enum` example to use a union, you will see the same behavior and error message.

```python
example3: list[tuple[X | int, str]] = [(X.a, 0)]
```

I realize the Python invariance rules and inference behaviors can be confusing. I don't mind providing explanations, but please trust me when I say that this is not a bug. You will find that mypy emits errors in the same places. I think pyright's error messages are more complete and understandable than mypy's because they explain that the type compatibility issue is related to the invariance of the list type parameter.
